splicemachine/spliceengine
# DB-12085 Fix isQualifier() for index conglomerates

## Short Description
This change prevents predicates that cannot be used to limit a table scan from being pushed down to base table wrongly and used a scan qualifier.

## Long Description
Before this fix, `isQualifier()` (`RelationalOperator` interface method) implementations simply checks if the column reference operand of this relational operator refers to a column of `optTable`. Depending on the type of relational operator, there might be other checks. But this is a necessary part of all implementations.

However, this check is not sufficient in case we want to push this predicate down to scan. For an index conglomerate, the column reference must refer to one of the index columns, too. This check is currently missing.

Note that `isQualifier()` is used in two cases, which have slightly different semantics:
1. check if a predicate can be used as a hash key
2. check if a predicate can be used as a scan qualifier

In case 1, the conglomerate columns check is not needed because we assume a predicate is on the level of a join. Down to the subtree, it can be an index scan + index lookup. But eventually, we have access to all columns referenced by predicates. Point is, we are not going to push the predicate down.

In case 2, we care about if it's possible to actually push this predicate down to the scan. In this case, the conglomerate columns check is necessary.

The two cases above are distinguished by the `forPush` argument.

One may wonder in case 2, how simple queries would even work without this check because it's very fundamental. Answer is, in `orderUsefulPredicates()`, `isQualifier()` is mostly called under a guard that the current conglomerate is the table conglomerate. If it's an index conglomerate, the code path calls `isIndexUseful()`.

But there is a missing piece. In case of a disjunction of predicates, e.g., `col1 = 2 or col2= 3`, we call `Predicate.isPushableOrClause()`, which in turns calls `isQualifier()` on each term of the disjunction without considering the current conglomerate. Consequently, the whole or-list can be pushed down to an index scan as a qualifier. However, some columns used in the predicate are not accessible during the scan, causing an ArrayIndexOutOfBound exception at run time.

The fix is to add conglomerate columns check for case 2 for index conglomerates. 

## How to test
Setup:
```
create table TA (c1 char(36));
create table TB (c2 char(36), s char(1), z char(1), hsp char(2), f char(1), pfnr char(10));
create index XPFHSPF on TB (hsp, f, s, z, c2);
```

Test query:
```
select * from --splice-properties joinOrder=fixed
 TA, TB --splice-properties index=XPFHSPF, joinStrategy=sortmerge
 where TA.c1 = TB.c2
   AND TB.s = '3' AND TB.z <> 'S' AND TB.hsp = 'LE'
   AND NOT (TB.f = 'L' AND TB.pfnr = '12345678  ')
```

Before this fix, running the test query would throw `ArrayIndexOutOfBound` exception. Note that explain this query returns a plan without any error. However, this plan is wrong because it takes `TB.f <> 'L' OR TB.pfnr <> '12345678  '` as a qualifier of index scan on `XPFHSPF`.

After this fix, the query runs correctly. Also, the explained plan should put `TB.f <> 'L' OR TB.pfnr <> '12345678  '` at the `ProjectRestrict` step on top of `IndexLookup`.
